renormalizer/mps/mp.py

```diff
@@ -645,11 +603,68 @@ def _update_mps(self, cstruct, cidx, qnbigl, qnbigr, Mmax, percent=0):
 
         # step 1: get the selected U, S, V
         if type(cstruct) is not list:
-            # SVD method
-            # full_matrices = True here to enable increase the bond dimension
-            Uset, SUset, qnlnew, Vset, SVset, qnrnew = svd_qn.svd_qn(
-                asnumpy(cstruct), qnbigl, qnbigr, self.qntot, system=system
-            )
+            if self.compress_config.ofs is None:
+                # SVD method
+                # full_matrices = True here to enable increase the bond dimension
+                Uset, SUset, qnlnew, Vset, SVset, qnrnew = svd_qn.svd_qn(
+                    asnumpy(cstruct), qnbigl, qnbigr, self.qntot, system=system
+                )
+            else:
+                if isinstance(self.model, HolsteinModel):
+                    # the HolsteinModel class methods are incompatible with OFS
+                    raise NotImplementedError("Can't perform OFS on Holstein model")
+
+                qnbigl1, qnbigr1 = qnbigl, qnbigr
+                Uset1, SUset1, qnlnew1, Vset1, SVset1, qnrnew1 = svd_qn.svd_qn(
+                    asnumpy(cstruct), qnbigl1, qnbigr1, self.qntot, system=system
+                )
+                qnbigl2, qnbigr2, _ = self._get_big_qn(cidx, swap=True)
+                if cstruct.ndim == 4:
+                    cstruct2 = asnumpy(cstruct).transpose(0, 2, 1, 3)
+                else:
+                    assert cstruct.ndim == 6
+                    cstruct2 = asnumpy(cstruct).transpose(0, 3, 4, 1, 2, 5)
+                if self.compress_config.ofs_swap_jw:
+                    assert cstruct2.ndim == 4
+                    cstruct2 = cstruct2.copy()
+                    cstruct2[:, 1, 1, :] = -cstruct2[:, 1, 1, :]
```

Comment:
1. This phase is assuming that |0>/|up> is unoccupied, |1>/|down> is occupied. This choice depends on the actual Jordan-Wigner transformation we use. I think it is better to add some assert or at least some comment somewhere.
2.  why is the cstruct2  copied?

Reply:
1. OK. 
2. because it is possible `cstruct2` shares the same underlying data with `cstruct`. Makes a copy can avoid overwrite `cstruct`